src/operations/execute_operation.ts

```diff
@@ -40,25 +40,31 @@ export function executeOperation<
   T extends OperationBase<TOptions, TResult>,
   TOptions = OptionsFromOperation<T>,
   TResult = ResultTypeFromOperation<T>
->(topology: Topology, operation: T): Promise<TResult>;
+>(topology: Topology | undefined, operation: T): Promise<TResult>;
 export function executeOperation<
   T extends OperationBase<TOptions, TResult>,
   TOptions = OptionsFromOperation<T>,
   TResult = ResultTypeFromOperation<T>
->(topology: Topology, operation: T, callback: Callback<TResult>): void;
+>(topology: Topology | undefined, operation: T, callback: Callback<TResult>): void;
 export function executeOperation<
   T extends OperationBase<TOptions, TResult>,
   TOptions = OptionsFromOperation<T>,
   TResult = ResultTypeFromOperation<T>
->(topology: Topology, operation: T, callback?: Callback<TResult>): Promise<TResult> | void;
+>(
+  topology: Topology | undefined,
+  operation: T,
+  callback?: Callback<TResult>
+): Promise<TResult> | void;
 export function executeOperation<
   T extends OperationBase<TOptions, TResult>,
   TOptions = OptionsFromOperation<T>,
   TResult = ResultTypeFromOperation<T>
->(topology: Topology, operation: T, callback?: Callback<TResult>): Promise<TResult> | void {
-  if (topology == null) {
-    throw new TypeError('This method requires a valid topology instance');
-  }
+>(
+  topology: Topology | undefined,
+  operation: T,
+  callback?: Callback<TResult>
+): Promise<TResult> | void {
+  if (!topology) throw new MongoClientClosedError();
```

Comment:
`getTopology` should be changed to always return a `Topology`, not `Topology | undefined` - that method will throw if one is not found. Likewise, you can remove this `topology == null` check since we now validate the presence of a `Topology` in `getTopology`